lib/internal/modules/esm/initialize_import_meta.js

```diff
@@ -47,6 +53,36 @@ function initializeImportMeta(meta, context, loader) {
   return meta;
 }
 
+/**
+ * Inspect the module URL and convert it to `__dirname` and `__filename`
+ * references if it does not point to a network module.
+ * @param {string} url
+ * @returns {{__dirname?: string, __filename?: string}}
+ */
+function resolveModuleMeta(url) {
+  if (url.slice(0, 7) !== 'file://') {
+    // These only make sense for locally loaded modules,
+    // i.e. network modules are not supported.
+    return { __dirname: undefined, __filename: undefined };
+  }
+
+  // We return an object with getters to reduce memory usage in modules that
+  // do not use these properties, i.e. most modules don't need the values
+  // to be resolved until the first time the property is accessed.
+  let filePath;
+  let dirPath;
+  return {
+    get __dirname() {
+      return dirname(filePath ??= fileURLToPath(url));
+    },
+    get __filename() {
+      filePath ??= fileURLToPath(url);
+      dirPath ??= dirname(filePath);
+      return filePath;
```

Comment:
What's the point of using getters here? It's going to be evaluated anyway.

Reply:
See https://github.com/nodejs/node/pull/48740#discussion_r1260338571

Reply:
I don't get it though, the getters are evaluated right away, there's no point in keeping them or am I missing something?

Reply:
I concur. I figured I was just missing something and went with the suggestion.

Reply:
I'm thinking on how I would implement the original ask by @bmeck without extraneous changes to the existing test. Originally, I wanted to define the properties on `meta` via `Object.defineProperties` so that I could make them non-writeable and non-configurable. But that violates the test https://github.com/nodejs/node/blob/f9fee9a8f41f9d6e04047fb71e80c95b50c927ed/test/es-module/test-esm-import-meta.mjs#L10-L17

Further, wrapping up the `moduleMeta` access via the same manner runs into the same issue with the added bonus of `get()` in the offending list of properties.

I'm open to suggestions on how to satisfy the ask.

Reply:
I think https://tc39.es/ecma262/#sec-hostgetimportmetaproperties stipulates that we cannot have a getter, it has to be a value.

Reply:
<details>
<summary>Sold!</summary>

![sold](https://user-images.githubusercontent.com/321201/253351213-c31ee551-5db3-46ea-a7e1-cecc7a3995b9.gif)
</details>

Reply:
@aduh95 this is incorrect per https://tc39.es/ecma262/#sec-hostfinalizeimportmeta